policies/policies.go

```diff
@@ -0,0 +1,301 @@
+			yumRepos = append(yumRepos, r)
+			continue
+		}
+		if r := repo.GetPackageRepository().GetZypper(); r != nil {
+			zypperRepos = append(zypperRepos, r)
+			continue
+		}
+	}
+
+	var aptInstalled, aptRemoved, aptUpdated []*osconfigpb.Package
+	var yumInstalled, yumRemoved, yumUpdated []*osconfigpb.Package
+	var zypperInstalled, zypperRemoved, zypperUpdated []*osconfigpb.Package
+	var gooInstalled, gooRemoved, gooUpdated []*osconfigpb.Package
+	for _, pkg := range res.GetPackages() {
+		switch pkg.GetPackage().GetManager() {
+		case osconfigpb.Package_ANY:
+			switch pkg.GetPackage().GetDesiredState() {
+			case osconfigpb.DesiredState_INSTALLED:
+				gooInstalled = append(gooInstalled, pkg.GetPackage())
+				aptInstalled = append(aptInstalled, pkg.GetPackage())
+				yumInstalled = append(yumInstalled, pkg.GetPackage())
+				zypperInstalled = append(zypperInstalled, pkg.GetPackage())
+			case osconfigpb.DesiredState_REMOVED:
+				gooRemoved = append(gooRemoved, pkg.GetPackage())
+				aptRemoved = append(aptRemoved, pkg.GetPackage())
+				yumRemoved = append(yumRemoved, pkg.GetPackage())
+				zypperRemoved = append(zypperRemoved, pkg.GetPackage())
+			case osconfigpb.DesiredState_UPDATED:
+				gooUpdated = append(gooUpdated, pkg.GetPackage())
+				aptUpdated = append(aptUpdated, pkg.GetPackage())
+				yumUpdated = append(yumUpdated, pkg.GetPackage())
+				zypperUpdated = append(zypperUpdated, pkg.GetPackage())
+			}
+		case osconfigpb.Package_GOO:
+			switch pkg.GetPackage().GetDesiredState() {
+			case osconfigpb.DesiredState_INSTALLED:
+				gooInstalled = append(gooInstalled, pkg.GetPackage())
+			case osconfigpb.DesiredState_REMOVED:
+				gooRemoved = append(gooRemoved, pkg.GetPackage())
+			case osconfigpb.DesiredState_UPDATED:
+				gooUpdated = append(gooUpdated, pkg.GetPackage())
+			}
+		case osconfigpb.Package_APT:
+			switch pkg.GetPackage().GetDesiredState() {
+			case osconfigpb.DesiredState_INSTALLED:
+				aptInstalled = append(aptInstalled, pkg.GetPackage())
+			case osconfigpb.DesiredState_REMOVED:
+				aptRemoved = append(aptRemoved, pkg.GetPackage())
+			case osconfigpb.DesiredState_UPDATED:
+				aptUpdated = append(aptUpdated, pkg.GetPackage())
+			}
+		case osconfigpb.Package_YUM:
+			switch pkg.GetPackage().GetDesiredState() {
+			case osconfigpb.DesiredState_INSTALLED:
+				yumInstalled = append(yumInstalled, pkg.GetPackage())
+			case osconfigpb.DesiredState_REMOVED:
+				yumRemoved = append(yumRemoved, pkg.GetPackage())
+			case osconfigpb.DesiredState_UPDATED:
+				yumUpdated = append(yumUpdated, pkg.GetPackage())
+			}
+		case osconfigpb.Package_ZYPPER:
+			switch pkg.GetPackage().GetDesiredState() {
+			case osconfigpb.DesiredState_INSTALLED:
+				zypperInstalled = append(zypperInstalled, pkg.GetPackage())
+			case osconfigpb.DesiredState_REMOVED:
+				zypperRemoved = append(zypperRemoved, pkg.GetPackage())
+			case osconfigpb.DesiredState_UPDATED:
+				zypperUpdated = append(zypperUpdated, pkg.GetPackage())
+			}
```

Comment:
here too I think you can factor out the block of these case statements into a single function call. Alternatively, you could create a new interface with some functions that work for each package manager type.

Reply:
I did some work factoring this out, I'll want to do some more and add better tests but that's probably best left for a future PR